### PR TITLE
Cucumber is added to the Project, please check.

### PR DESCRIPTION
**Pull request by yulia-borisova**
Hi Artem,
Project  with Cucumber is committed ,could you please review

### CODE REVIEW

File: PageObjectPattern/src/main/java/shopping/LoginPage.java [hunk 1]
```diff
@@ -26,6 +27,17 @@ public LoginPage(WebDriver driver) {
     @FindBy(id = "SubmitCreate")
     private WebElement submitRegisterButton;
 
+    // Element on the page to Reset Password
+
+    @FindBy(how = How.LINK_TEXT, using = "Forgot your password?")
+    WebElement forgotPass;
+
+    // Method used to reset pass and navigate to a new page
+    public ResetPasswordPage clickForgotPassword() {
```
Comment (alaktionov) [inline PageObjectPattern/src/main/java/shopping/LoginPage.java]: clickForgotPassword -> clickForgotPasswordLink

File: PageObjectPattern/src/main/java/shopping/ResetPasswordPage.java [hunk 2]
```diff
@@ -16,15 +16,15 @@ public ResetPasswordPage(WebDriver driver) {
     private WebElement email;
 
     @FindBy(xpath = "//*[@id=\"form_forgotpassword\"]//button")
-    private WebElement submitButton;
+    private WebElement retrievPasswordButton;
```
Comment (alaktionov) [inline PageObjectPattern/src/main/java/shopping/ResetPasswordPage.java]: retrievPasswordButton -> retrievePasswordButton

File: PageObjectPattern/src/main/java/shopping/ResetPasswordPage.java [hunk 3]
```diff
@@ -16,15 +16,15 @@ public ResetPasswordPage(WebDriver driver) {
     private WebElement email;
 
     @FindBy(xpath = "//*[@id=\"form_forgotpassword\"]//button")
```
Comment (alaktionov) [inline PageObjectPattern/src/main/java/shopping/ResetPasswordPage.java]: xpath can be changed to "//*[@id='form_forgotpassword']//button"

File: PageObjectPattern/src/main/java/shopping/SignInFormPage.java [hunk 4]
```diff
@@ -80,6 +80,11 @@ public void enterAddress(String addressFirstName, String addressLastName, String
         this.aliasAddressInput.sendKeys(aliasAddress);
     }
 
+    public String getUrl() {
+        String url = driver.getCurrentUrl();
```
Comment (alaktionov) [inline PageObjectPattern/src/main/java/shopping/SignInFormPage.java]: Local variable url is not needed here, you could just use return driver.getCurrentUrl()

Besides, maybe this method is redundant.

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/LoginSteps.java [hunk 5]
```diff
@@ -24,21 +24,21 @@ public void user_is_on_login_page() {
     }
 
     @When("User inputs valid {word} and {word}")
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/ShoppingStepDifinitions/LoginSteps.java:26]: {word} and {word} can be replaced with {email} and {password}

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepOne.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+public class RegistrationStepOne {
+    WebDriver driver;
+    HomePage homePage;
+    LoginPage loginPage;
+    SignInFormPage formPage;
+
+    @Given("New User is on Home Page")
+    public void getHomePage() {
+        System.out.println("Step 1: User is on login page");
+        System.setProperty("webdriver.chrome.driver", "src\\test\\resources\\drivers\\chromedriver.exe");
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepOne.java:24]: You already have this property set in getLoginPage() method. Think of extracting this to one common method

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepTwo.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    @Given("User is on SignIn Form Page")
+    public void getSignInForm() {
+        System.out.println("Step1: New User is on Registration Form page");
+
+    }
+
+    @When("^User inputs (.+), (.+), (.+), (.+),(.+), (.+),(.+), (.+), (.+), (.+), (.+), (.+)$")
+    public void inputUserData(String firstname, String lastname, String password, String day, String month, String year,
+            String address, String city, String state, String code, String phone, String alias) {
+        System.out.println("Step2: User inputs data");
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepTwo.java:17]: Missing method for actual input of the data

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepTwo.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    @And("Clicks Register button")
+    public void clickRegisterButton() {
+        System.out.println("Step3: User clicks register button");
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepTwo.java:22]: No method for clicking the button

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepTwo.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    @Then("User is navigated to My Account Page")
+    public void user_is_navigated_to_my_account_page() {
+        System.out.println("Step4: User is navitated to My Account Page");
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepTwo.java:27]: No method for varification

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/ResetPasswordSteps.java [hunk 10]
```diff
@@ -1,35 +1,66 @@
...
 public class ResetPasswordSteps {
+    WebDriver driver;
+    HomePage homePage;
+    LoginPage loginPage;
+    ResetPasswordPage resetPassPage;
+
     @Given("User is on Home page")
-    public void user_is_on_home_page() {
-        System.out.println("Step 1: User is on home page");
+    public void getHomePage() {
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/ShoppingStepDifinitions/ResetPasswordSteps.java:24]: This method already define in a different class, please check

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/TestRunner.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package ShoppingStepDifinitions;
+
+import org.junit.runner.RunWith;
+
+import io.cucumber.junit.Cucumber;
+import io.cucumber.junit.CucumberOptions;
+
+//To run Feature test suits with specified tags & save reports in different formats
+@RunWith(Cucumber.class)
+@CucumberOptions(features = "src\\test\\resources\\Features", glue = { "ShoppingStepDifinitions" }, plugin = { "pretty",
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/ShoppingStepDifinitions/TestRunner.java]: Path to features can be replaced with "classpath:features". Also rename Features -> features

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/TestRunner.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+        "json:target/JSONReports", "junit:target/JUnitReports" }, tags = "@smoketest")
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/ShoppingStepDifinitions/TestRunner.java:11]: Think of how you could generate Serenity report

File: PageObjectPattern/src/test/resources/Features/registrationStepOne.feature [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+  Scenario Outline: User creates a new account
+    Given New User is on Home Page
+    And Clicks SignIn
+    When New User Inputs <email>
+    And Clicks Create an Account
+    Then User is redirected to SignIn Form Page
+
+    Examples: 
+      | email                      |
```
Comment (alaktionov) [inline PageObjectPattern/src/test/resources/Features/registrationStepOne.feature]: Align the table :)

File: PageObjectPattern/src/main/java/shopping/LoginPage.java (hunk 1 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/main/java/shopping/LoginPage.java]: done

File: PageObjectPattern/src/main/java/shopping/ResetPasswordPage.java (hunk 2 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/main/java/shopping/ResetPasswordPage.java]: done

File: PageObjectPattern/src/main/java/shopping/ResetPasswordPage.java (hunk 3 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/main/java/shopping/ResetPasswordPage.java]: done

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepTwo.java (hunk 9 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/ShoppingStepDifinitions/RegistrationStepTwo.java:27]: done

File: PageObjectPattern/src/main/java/shopping/SignInFormPage.java (hunk 4 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/main/java/shopping/SignInFormPage.java]: done

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/ResetPasswordSteps.java (hunk 10 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/ShoppingStepDifinitions/ResetPasswordSteps.java:24]: done

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/TestRunner.java (hunk 11 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/ShoppingStepDifinitions/TestRunner.java]: done

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/TestRunner.java (hunk 12 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/ShoppingStepDifinitions/TestRunner.java:11]: will be done in the next project

File: PageObjectPattern/src/test/resources/Features/registrationStepOne.feature (hunk 13 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/resources/Features/registrationStepOne.feature]: done

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/LoginSteps.java (hunk 5 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/ShoppingStepDifinitions/LoginSteps.java:26]: getting exception in this case  "Could not create a cucumber expression for 'User inputs valid {email} and {password}'

**Comment by yulia-borisova** [comment]
Hi,
please review .
Thanks!

File: PageObjectPattern/src/test/java/ShoppingStepDifinitions/LoginSteps.java (hunk 5 above)
Comment (alaktionov) [reply PageObjectPattern/src/test/java/ShoppingStepDifinitions/LoginSteps.java:26]: You could also try just 'User inputs valid {} and {}'

**Review by alaktionov — APPROVED** [review]

**Comment by yulia-borisova** [comment]
Hi Artem,
I've added Serenity to the existing project and created new folders : serenity.pages, serenitycucumber.runner, serenityCucumberStepDefinitions , serenitycucumber.features.
Could you please review.

File: PageObjectPattern/pom.xml [hunk 14]
```diff
@@ -1,54 +1,180 @@
...
+		<serenity.version>2.2.0</serenity.version>
+		<serenity.maven.version>2.2.0</serenity.maven.version>
+		<serenity.cucumber.version>2.2.0</serenity.cucumber.version>
+		<cucumber.version>5.5.0</cucumber.version>
+		<encoding>UTF-8</encoding>
+		<tags></tags>
+		<parallel.tests>4</parallel.tests>
+		<webdriver.base.url></webdriver.base.url>
+	</properties>
+	<repositories>
```
Comment (alaktionov) [inline PageObjectPattern/pom.xml]: Do we really need these two repositories here?

File: PageObjectPattern/pom.xml [hunk 15]
```diff
@@ -1,54 +1,180 @@
...
 		<dependency>
-			<groupId>io.cucumber</groupId>
-			<artifactId>cucumber-junit</artifactId>
-			<version>6.4.0</version>
+			<groupId>net.serenity-bdd</groupId>
+			<artifactId>serenity-cucumber5</artifactId>
+			<version>${serenity.cucumber.version}</version>
 			<scope>test</scope>
 		</dependency>
 		<dependency>
```
Comment (alaktionov) [inline PageObjectPattern/pom.xml:86]: No need to use all 3 libraries for assertions (junit, assertj, hamcrest), only one of them can be used. I prefer `org.assertj` personally

File: PageObjectPattern/serenity.properties [hunk 16]
```diff
@@ -0,0 +1,3 @@
+serenity.project.name=PageObjectPattern
+drivers.windows.webdriver.chrome.driver = src/test/resources/drivers/chromedriver.exe
```
Comment (alaktionov) [inline PageObjectPattern/serenity.properties]: Spaces are redundant before and after `=`

File: PageObjectPattern/src/test/java/serenity/googletest/GoogleSearchTest.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.openqa.selenium.By;
+import org.openqa.selenium.Keys;
+import org.openqa.selenium.WebDriver;
+
+import net.serenitybdd.junit.runners.SerenityRunner;
+import net.thucydides.core.annotations.Managed;
+@RunWith(SerenityRunner.class)
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenity/googletest/GoogleSearchTest.java:14]: Put a break line before @RunWith

File: PageObjectPattern/src/test/java/serenity/googletest/GoogleSearchTest.java [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+    
+    @Managed
+    WebDriver driver;
+    
+    @Test
+    public void search() {
+        driver.get("https://www.google.com/");
+        driver.findElement(By.name("q")).sendKeys("serenity-bdd", Keys.ENTER);
+        assertTrue(("message: something went wrong"),driver.findElement(By.id("result-stats")).isDisplayed());
+
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenity/googletest/GoogleSearchTest.java]: Redundant break line

File: PageObjectPattern/src/test/java/serenity/pages/ForgotPasswordPage.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
+package serenity.pages;
+
+import org.openqa.selenium.By;
+import net.serenitybdd.core.steps.UIInteractionSteps;
+import net.serenitybdd.screenplay.ensure.BooleanEnsure;
+import net.serenitybdd.screenplay.ensure.Ensure;
+import net.thucydides.core.annotations.Step;
+
+//Elements and Methods on Forgot password page
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenity/pages/ForgotPasswordPage.java]: I remember I left a comment to your previous PR about comments. You can either remove them from all classes or if you need them, just enclose into /** */ comments block

File: PageObjectPattern/src/test/java/serenity/pages/LoginPage.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+    By signInButton = By.id("SubmitLogin");
+    By forgotPasswordLink = By.xpath("//*[@title=\"Recover your forgotten password\"]");
+
+    // Methods used on login page
+    @Step("Open Login page")
+    public void openLoginPage() {
+        this.openUrl("http://automationpractice.com/index.php?controller=authentication&back=my-account");
+    }
+
+    @Step("User provides credentials")
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenity/pages/LoginPage.java:21]: There is also a possibility to show values of the parameters on the report if you use for instance:
`User provides credentials - login {0}, password {1}`

File: PageObjectPattern/src/test/java/serenity/pages/MyAccountPage.java [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+import net.thucydides.core.annotations.Step;
+
+public class MyAccountPage extends UIInteractionSteps {
+    MainPage page;
+
+    By myInfo = By.xpath("//*[@class='myaccount-link-list']");
+
+    @Step("Check successfull login")
+    public BooleanEnsure loginCheck() {
+
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenity/pages/MyAccountPage.java]: This break line is redundant

File: PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/LoginStepDefinitions.java [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+//Cucumber glue code, using regex
+public class LoginStepDefinitions {
+
+    @Steps
+    LoginPage loginPage;
+
+    @Steps
+    MyAccountPage userPage;
+
+    @Given("^Unauthenticated User is on Login Page$")
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/LoginStepDefinitions.java:17]: Cucumber will understand you step if you just use:
`Unauthenticated User is on Login Page`

File: PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/LoginStepDefinitions.java [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+    public void getPage() {
+        System.out.println("Step 1");
+        loginPage.openLoginPage();
+    }
+
+    @When("^User inserts (.+) , (.+)$")
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/LoginStepDefinitions.java:23]: This step definition can be simplified to the following:
`User inserts {word} , {word}` or `User inserts {} , {}`

Other useful expressions https://cucumber.io/docs/cucumber/cucumber-expressions/

File: PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/RetrievePassStepDefinitions.java [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+    public void getHomePage() {
+        System.out.println("Step 1");
+        homePage.openHomePage();
+    }
+
+    @When("User clicks SignIn button")
+    public void clickSignInButton() {
+        System.out.println("Step 2");
+        homePage.clickSignInButton();
+
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/RetrievePassStepDefinitions.java]: Redundant break line

**Review by alaktionov — COMMENTED** [review]
Looks good Julia, although I left some remarks. Please check them out. Additionally make sure that Serenity report is successfully generated. You could send me the report once all comments are fixed :)

File: PageObjectPattern/pom.xml (hunk 14 above)
Comment (yulia-borisova) [reply PageObjectPattern/pom.xml]: done

File: PageObjectPattern/pom.xml (hunk 15 above)
Comment (yulia-borisova) [reply PageObjectPattern/pom.xml:86]: done

File: PageObjectPattern/serenity.properties (hunk 16 above)
Comment (yulia-borisova) [reply PageObjectPattern/serenity.properties]: done

File: PageObjectPattern/src/test/java/serenity/googletest/GoogleSearchTest.java (hunk 17 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/serenity/googletest/GoogleSearchTest.java:14]: done

File: PageObjectPattern/src/test/java/serenity/googletest/GoogleSearchTest.java (hunk 18 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/serenity/googletest/GoogleSearchTest.java]: done

File: PageObjectPattern/src/test/java/serenity/pages/MyAccountPage.java (hunk 21 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/serenity/pages/MyAccountPage.java]: done

File: PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/LoginStepDefinitions.java (hunk 22 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/LoginStepDefinitions.java:17]: got it, thanks

File: PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/LoginStepDefinitions.java (hunk 23 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/LoginStepDefinitions.java:23]: just tried using regex)

File: PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/RetrievePassStepDefinitions.java (hunk 24 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/serenityCucumberStepDefinitions/RetrievePassStepDefinitions.java]: done

File: PageObjectPattern/src/test/java/serenity/pages/ForgotPasswordPage.java (hunk 19 above)
Comment (yulia-borisova) [reply PageObjectPattern/src/test/java/serenity/pages/ForgotPasswordPage.java]: done

**Comment by yulia-borisova** [comment]
All comments resolved + summary report is added.
Could you please review.

Many thanks ,
Yulia

**Comment by yulia-borisova** [comment]
Maven postman dependency added.

File: PageObjectPattern/src/test/java/serenitycucumber/runner/SendEmailReport.java [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+
+        MultiPartEmail emailAttachment = new MultiPartEmail();
+        emailAttachment.setHostName("smtp.gmail.com");
+        emailAttachment.setSmtpPort(465);
+        emailAttachment.setAuthenticator(new DefaultAuthenticator("yulia.borisova.tr@gmail.com", "Qa123456"));
+        emailAttachment.setSSLOnConnect(true);
+        emailAttachment.setFrom("yulia.borisova.tr@gmail.com");
+        emailAttachment.setSubject("TestMail");
+        emailAttachment.setMsg("Hi Artem! Sending you my first Serenity report. Have a nice day:)");
+        emailAttachment.addTo("yulia.borisowa.qa@gmail.com");
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenitycucumber/runner/SendEmailReport.java:30]: Interesting, should it be my email? laktionov_artem@ukr.net :)

File: PageObjectPattern/src/test/java/serenitycucumber/runner/SendEmailReport.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
+package serenitycucumber.runner;
+
+import org.apache.commons.mail.DefaultAuthenticator;
+import org.apache.commons.mail.EmailAttachment;
+import org.apache.commons.mail.EmailException;
+import org.apache.commons.mail.MultiPartEmail;
+
+/*change settings in Google email: Google Account >> Security >> Less Secure App access >> turn it ON*/
```
Comment (alaktionov) [inline PageObjectPattern/src/test/java/serenitycucumber/runner/SendEmailReport.java:8]:
```suggestion
/**
Change settings in Google email: Google Account >> Security >> Less Secure App access >> turn it ON
*/
```

File: PageObjectPattern/.classpath [hunk 27]
```diff
@@ -24,7 +24,7 @@
 			<attribute name="maven.pomderived" value="true"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/J2SE-1.5">
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8">
```
Comment (alaktionov) [inline PageObjectPattern/.classpath:27]: No need to commit these changes to the repository, this file should be in .gitignore

File: PageObjectPattern/.settings/org.eclipse.jdt.core.prefs [hunk 28]
```diff
@@ -1,8 +1,8 @@
...
-org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.5
-org.eclipse.jdt.core.compiler.compliance=1.5
+org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.8
+org.eclipse.jdt.core.compiler.compliance=1.8
 org.eclipse.jdt.core.compiler.problem.enablePreviewFeatures=disabled
 org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
 org.eclipse.jdt.core.compiler.problem.reportPreviewFeatures=ignore
 org.eclipse.jdt.core.compiler.release=disabled
-org.eclipse.jdt.core.compiler.source=1.5
+org.eclipse.jdt.core.compiler.source=1.8
```
Comment (alaktionov) [inline PageObjectPattern/.settings/org.eclipse.jdt.core.prefs:8]: No need to commit these changes to the repository, this file should be in .gitignore

**Review by alaktionov — APPROVED** [review]

**Pull request merged**